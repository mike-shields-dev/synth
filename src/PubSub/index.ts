export { ControlChangeSubscriber, publishControlChange } from './ControlChange';
export type { ControlChange } from './ControlChange/types';
export { MidiNoteOnSubscriber, publishMidiNoteOn } from './MidiNoteOn';
export type { MidiNoteOn } from './MidiNoteOn/types';
export { NoteOffSubscriber, publishNoteOff } from './NoteOff';
export type { NoteOff } from './NoteOff/types';

