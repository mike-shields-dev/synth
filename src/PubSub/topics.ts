const
    MIDI = 'midi',
    MIDI_CC = 'midi.controlChange',
    MIDI_NOTE_ON = 'midi.note.on',
    MIDI_NOTE_OFF = 'midi.note.off',
    UI_CC = 'ui.controlChange',
    FOCUS = 'focus';

export {
    MIDI, 
    MIDI_CC, 
    MIDI_NOTE_ON,
    MIDI_NOTE_OFF,
    UI_CC,
    FOCUS,
}