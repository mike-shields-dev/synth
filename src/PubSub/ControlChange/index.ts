import PubSub from 'pubsub-js';
import { CONTROL } from '../topics';
import { ControlChange, ControlChangeHandler } from './types';

function publishControlChange(data: ControlChange) {
  PubSub.publish(CONTROL, data);
}

class ControlChangeSubscriber {
  private subscription: string;

  constructor(controlChangeHandler: ControlChangeHandler) {
    this.subscription = PubSub.subscribe(
      CONTROL,
      controlChangeHandler,
    );
  }

  public unsubscribe() {
    PubSub.unsubscribe(this.subscription);
  }
}

export {
  publishControlChange,
  ControlChangeSubscriber
};

