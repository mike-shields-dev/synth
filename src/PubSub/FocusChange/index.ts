import PubSub from "pubsub-js";
import { FOCUS } from '../topics';
import { FocusChangeHandler } from "./types";

function publishFocusChange(data: string) {
    PubSub.publish(FOCUS, data);
}

class FocusChangeSubscriber {
    private subscription: string;

    constructor(focusHandler: FocusChangeHandler) {
        this.subscription = PubSub.subscribe(
            FOCUS,
            focusHandler,
        );
    }

    public unsubscribe() {
        PubSub.unsubscribe(this.subscription);
    }
}

export {
    publishFocusChange,
    FocusChangeSubscriber
};
