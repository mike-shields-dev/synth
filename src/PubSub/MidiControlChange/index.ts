import PubSub from 'pubsub-js';
import { MIDI_CC } from '../topics';
import { MidiControlChange, MidiControlChangeHandler } from './types';

function publishMidiControlChange(data: MidiControlChange) {
  PubSub.publish(MIDI_CC, data);
}

class MidiControlChangeSubscriber {
  private subscription: string;

  constructor(controlChangeHandler: MidiControlChangeHandler) {
    this.subscription = PubSub.subscribe(
      MIDI_CC,
      controlChangeHandler,
    );
  }

  public unsubscribe() {
    PubSub.unsubscribe(this.subscription);
  }
}

export {
  publishMidiControlChange,
  MidiControlChangeSubscriber
};

