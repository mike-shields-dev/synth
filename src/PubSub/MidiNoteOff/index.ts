import PubSub from "pubsub-js";
import { MIDI_NOTE_OFF } from "../topics";
import { MidiNoteOff, MidiNoteOffHandler } from './types';

function publishMidiNoteOff(payload: MidiNoteOff) {
    PubSub.publish(MIDI_NOTE_OFF, payload)
}

class MidiNoteOffSubscriber {
    private subscription: string;

    constructor(noteOffHandler: MidiNoteOffHandler) {
        this.subscription = PubSub.subscribe(
            MIDI_NOTE_OFF,
            noteOffHandler,
        )
    }

    public unsubscribe() {
        PubSub.unsubscribe(this.subscription);
    };
}

export { publishMidiNoteOff, MidiNoteOffSubscriber };
