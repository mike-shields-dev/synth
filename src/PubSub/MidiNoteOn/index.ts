import PubSub from "pubsub-js";
import { MIDI_NOTE_ON } from "../topics";
import { MidiNoteOn, MidiNoteOnHandler } from "./types";

function publishMidiNoteOn(payload: MidiNoteOn) {
    PubSub.publish(MIDI_NOTE_ON, payload)
}

class MidiNoteOnSubscriber {
    private subscription: any;

    constructor(noteOnHandler: MidiNoteOnHandler) {
        this.subscription = PubSub.subscribe(
            MIDI_NOTE_ON,
            noteOnHandler,
        )
    }

    public unsubscribe() {
        PubSub.unsubscribe(this.subscription);
    };
}

export { publishMidiNoteOn, MidiNoteOnSubscriber };
