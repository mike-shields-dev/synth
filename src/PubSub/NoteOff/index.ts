import PubSub from "pubsub-js";
import { MIDI_NOTE_OFF } from "../topics";
import { NoteOff, NoteOffHandler } from './types';

function publishNoteOff(data: NoteOff) {
    PubSub.publish(MIDI_NOTE_OFF, data)
}

class NoteOffSubscriber {
    private subscription: string;

    constructor(noteOffHandler: NoteOffHandler) {
        this.subscription = PubSub.subscribe(
            MIDI_NOTE_OFF,
            noteOffHandler,
        )
    }

    public unsubscribe() {
        PubSub.unsubscribe(this.subscription);
    };
}

export { publishNoteOff, NoteOffSubscriber };
