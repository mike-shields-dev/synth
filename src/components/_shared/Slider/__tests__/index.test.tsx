import { render, screen, waitFor, fireEvent } from '@testing-library/react';
import { vi } from 'vitest';
import { Slider } from '..';
import { MIDI_CC } from '../../../../PubSub/topics';

const validProps = {
    controlChangeNumber: 70,
    group: 'groupName',
    initVal: 65,
    isFocused: true,
    parameter: 'parameter',
    scalers: {
        out: vi.fn((n) => n * 2),
    }
}

describe('Slider', () => {
    it('matches snapshot', () => {
        const { container } = render(<Slider {...validProps} />);

        expect(container.firstChild).toMatchSnapshot();
    });

    beforeEach(() => render(< Slider {...validProps} />));
    afterEach(() => vi.resetAllMocks());

    it('renders a slider with the give parameter name', () => {
        expect(screen.getByRole('slider', { name: /parameter/i })).toBeInTheDocument();
    });

    it('renders a slider with the given initial value', () => {
        expect(screen.getByRole('slider', { name: /Parameter/i })).toHaveValue(`${validProps.initVal}`);


    })

    it('renders a single range input', () => {
        expect(screen.getAllByRole('slider')).toHaveLength(1);
    });

    it('renders a range input with a min value of 0', () => {
        expect(screen.getByRole('slider')
            .getAttribute('min'))
            .toEqual('0');
    });

    it('renders a range input with a max value of 127', () => {
        expect(screen.getByRole('slider')
            .getAttribute('max'))
            .toEqual('127');
    });

    it('renders an output element', () => {
        expect(screen.getByRole('status')).toBeInTheDocument();
    });

    it('renders an output element associated with the input element', ()=> {
        expect(screen.getByRole('status')
            .getAttribute('for'))
            .toBe(`${validProps.group}-${validProps.parameter}`);
    });

    it(`updates slider's value when a MIDI Control Change is published with the correct payload`, async () => {
        const slider = screen.getByRole('slider', { name: /Parameter/i });
        
        const payload = {
            controlChangeNumber: validProps.controlChangeNumber,
            value: 99,
        };

        PubSub.publish(MIDI_CC, payload);
        
        await waitFor(() => {
            expect(slider).toHaveValue(`${payload.value}`)
        });
    });

    it(`updates sliders value on user input via the GUI`, async () => {
        const slider = screen.getByRole('slider', { name: /Parameter/i });
        const newVal = 99;

        fireEvent.change(slider, { target: { value: newVal } });

        await waitFor(() => {
            expect(slider).toHaveValue(`${newVal}`);
        });
    });

    it('displays the scaled value of the slider', async () => {
        const outputDisplay = screen.getByRole('status');
        
        expect(outputDisplay).toHaveValue(`${validProps.scalers.out(validProps.initVal)}`);
    });

    it('updates the displayed scaled value when a MIDI Control Change is published with the correct payload', async () => {
        const outputDisplay = screen.getByRole('status');
        const payload = {
            controlChangeNumber: validProps.controlChangeNumber,
            value: 99,
        };

        PubSub.publish(MIDI_CC, payload);

        await waitFor(() => {
            expect(outputDisplay).toHaveValue(`${validProps.scalers.out(validProps.initVal)}`);
        });
    });
});