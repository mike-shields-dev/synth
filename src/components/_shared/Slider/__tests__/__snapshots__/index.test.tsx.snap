// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Slider > matches snapshot 1`] = `
<div>
  <label
    for="groupName-parameter"
  >
    Parameter
  </label>
  <input
    id="groupName-parameter"
    max="127"
    min="0"
    step="0.05"
    type="range"
    value="65"
  />
  <output
    aria-label="Group Name parameter"
    for="groupName-parameter"
  >
    130.00
  </output>
</div>
`;
